### PR TITLE
Rework torrent411 plugin

### PR DESCRIPTION
**Pull request by gregaou**
I've just create a plugin for torrent411 french tracker based on actual urlrewrite_torrent411 plugin

### CODE REVIEW

File: flexget/plugins/search_t411.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                entry['torrent_seeds'] = torrent['seeders']
+                entry['torrent_leeches'] = torrent['leechers']
+                entry['search_sort'] = torrent_availability(entry['torrent_seeds'],
+                                                            entry['torrent_leeches'])
+                entry['content_size'] = torrent['size']
+                entry['download_auth'] = auth_handler
+                entries.add(entry)
+
+
+            return sorted(entries, reverse=True,
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: Shouldn't this be unindented one level? You return immediately after querying for the first search string.

File: flexget/plugins/search_t411.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        url_auth = T411_BASE_URL + "/auth"
+        db_session = Session()
+        account = db_session.query(t411Account).filter(
+            t411Account.username == username).first()
+        if account:
+            if account.expiry_time < datetime.now():
+                db_session.delete(account)
+                db_session.commit()
+            log.debug("Token found in db!")
+            return account.token
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: Is the token still valid after `expiry_time`?

File: flexget/plugins/search_t411.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+            if 'errors' in res_json.keys():
+                auth_handler = t411Auth(config.get('username'), config.get('password'), force_auth=True)
+                res_json = requests.get(T411_BASE_URL + url_search, auth=auth_handler).json()
+
+                if 'errors' in res_json.keys():
+                    raise plugin.PluginError(res['error'])
+
+            for torrent in res_json['torrents']:
+                entry = Entry()
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: I think it's bad practice to overwrite function arguments -- I know a few of the old search plugins do it, but new ones shouldn't. Python does not have local scopes like in C.

File: flexget/plugins/search_t411.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+class t411Auth(AuthBase):
+
+    """ Attaches HTTP Token Authentication to the given Request object."""
+
+    def get_auth(self):
+        url_auth = T411_BASE_URL + "/auth"
+        log.debug("Getting token from : %s ", url_auth)
+        params = {'username': self.username, 'password': self.password}
+        res = requests.post(url_auth, data=params).json()
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: You should probably wrap this in a try-except block.

File: flexget/plugins/search_t411.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            raise plugin.PluginError("Unable to get token ! (%s)" % res)
+
+        return res['token']
+
+    def update_token(self, db_session, token):
+        db_session.token = token
+        db_session.expiry_time = datetime.now() + timedelta(days=90)
+
+    def get_token(self):
+        db_session = Session()
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: If you use `with Session() as db_session:`, then you don't have to call commit().

File: flexget/plugins/search_t411.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+                                                          for c in sub_categories if c != 0])
+        entries = set()
+        auth_handler = t411Auth(config.get('username'), config.get('password'))
+        for search_string in entry.get('search_strings', [entry['title']]):
+            query = normalize_unicode(search_string)
+            url_search = ('/torrents/search/' +
+                          quote_plus(query.encode('utf-8')) +
+                          filter_url)
+
+            response = requests.get(T411_BASE_URL + url_search, auth=auth_handler)
```
Comment (cvium) [inline flexget/plugins/search_t411.py]: You should use `task.requests` instead and maybe also in the auth class.

File: flexget/plugins/search_t411.py (hunk 4 above)
Comment (cvium) [inline flexget/plugins/search_t411.py]: Requests will default to raising RequestExceptions on error status eg. 404.

**Comment by lildadou** [comment]
Hello, I making a [T411 plugin based on the rest api too.](https://github.com/lildadou/Flexget/tree/t411-rest). It has pretty cool features : proxy-cache, parser for categories & terms, lookup for torrent's details, 2 layers client (one is just REST client, the other is a proxy-cache above the rest client).

Are you interested to working together?
PS : on pourra parler en français en plus :)

**Comment by gregaou** [comment]
No problem, you can find me on #flexget irc channel with gregwar nickname.

**Comment by gazpachoking** [comment]
@gregaou @lildadou Did you guys work something out? Is this good to go?

**Comment by gregaou** [comment]
i have no news from @lildadou

**Comment by lildadou** [comment]
@gazpachoking My branch is still WIP. The main missed part is session authentication (http://discuss.flexget.com/t/session-for-plugins/69) for make torrent download effective.

@gregaou : Do you want to work on that? I'm currently working on the CLI.

**Comment by liiight** [comment]
This is no longer relevant right?

**Comment by gregaou** [comment]
Yep i close !

**Pull request closed without merging**